TNL-6700, TNL-6821 – Enable course sharing on web with initial eventing

Qubad786 opened the pull request:
## [TNL-6700](https://openedx.atlassian.net/browse/TNL-6700), [TNL-6821](https://openedx.atlassian.net/browse/TNL-6821)

### Description
This change will utilize `get_link_for_about_page` to retrieve course about url in order to share it on Twitter and Facebook, this course about url  can be either of following: 

1. `social_sharing_url`
2. `marketing_url`
3. link to LMS course about page

This also adds Eventing for Web Course Sharing.

**Event** 
`edx.course.share_clicked`

**Parameters**
`course_id`: Course key
`social_media_site`: FB/Twitter
`location`: Location where sharing attempt was made

**Description** 
Gets emitted when Social Media(i.e. FB/Twitter) share link is clicked. 

### Sandbox
- [x] https://course-sharing-on-lms-dashboard.sandbox.edx.org/dashboard
### Testing
- [x] Unit, integration, acceptance tests as appropriate

FYI: Tag anyone who might be interested in this PR here.

### Post-review
- [ ] Rebase and squash commits

Qubad786 commented:
sandbox: http://jenkins.edx.org:8080/view/Ansible/job/ansible-provision/11509/

Qubad786 commented:
FYI @muzaffaryousaf

muzaffaryousaf commented:
@Qubad786 Let's add tests for this change.

sstack22 commented:
@Qubad786  - this is looking great! I'll follow up with marketing to get thoughts on a default message for the share text.

Qubad786 commented:
@muzaffaryousaf tests are already there: https://github.com/edx/edx-platform/blob/master/common/test/acceptance/tests/lms/test_lms_dashboard.py#L151-L173

muzaffaryousaf commented:
@srpearce  FYI ^^

muzaffaryousaf commented:
> @Qubad786 - this is looking great! I'll follow up with marketing to get thoughts on a default message for the share text.

@sstack22 should we wait for this ?

muzaffaryousaf commented:
@Qubad786 This LGTM 👍 but wait for @sstack22 's comment for share text.

srpearce commented:
Adding @grantgoodmanedX, since he's covering this feature for the doc team.

Qubad786 commented:
@sstack22 please have a look, I've added eventing on share click, can be verified via browser console: https://course-sharing-on-lms-dashboard.sandbox.edx.org/dashboard

Also, who might be the right person to review the events from Analytics?

muzaffaryousaf commented on lms/static/js/course_sharing/course_sharing_events.js:
```diff
@@ -0,0 +1,29 @@
+
+            $(".action-facebook[data-course-id='" + courseId + "']").on('click', function(){
+                // Emit an event telling that the Facebook share link was clicked.
+                Logger.log('edx.course.sharing.share_clicked', {
```
@mulby Can you please review this event for course sharing feature.

muzaffaryousaf replied:
Also, we need your opinion on it's naming convention and either it should be research or bi ? (from previous ticket i think it will be bi event?).

muzaffaryousaf replied:
@mulby For testing you can use this sandbox https://course-sharing-on-lms-dashboard.sandbox.edx.org/login?next=/dashboard

muzaffaryousaf replied:
@sstack22  FYI ^^

mulby commented:
@muzaffaryousaf - I'm not able to see the sharing links on the sandbox. I logged in on using the staff user.

![screenshot from 2017-04-12 14-05-39](https://cloud.githubusercontent.com/assets/627177/24972341/2500049a-1f89-11e7-8c6a-fb9980293be5.png)

Another note, I don't think `get_link_for_about_page` does what we want it to. We need the link to contain the appropriate UTM parameters so that we can tell when the link is used. @sstack22 and I discussed this a while ago. Looking at the code, I don't see where those parameters are added here...

mulby commented:
@stroilova - can you quickly review these events for the data package? We don't need to emit them to the package, but that's the easiest way to instrument this code.

This is time-sensitive AFAIK.

mulby commented on lms/static/js/course_sharing/course_sharing_events.js:
```diff
@@ -0,0 +1,28 @@
+                // Emit an event telling that the Facebook share link was clicked.
+                Logger.log('edx.course.sharing.share_clicked', {
+                    course_id: courseId,
+                    social_media_site: 'Facebook',
```
I tend to prefer all lowercase tokens like: "facebook", "twitter" etc.

mulby commented on lms/static/js/course_sharing/course_sharing_events.js:
```diff
@@ -0,0 +1,28 @@
+                Logger.log('edx.course.sharing.share_clicked', {
+                    course_id: courseId,
+                    social_media_site: 'Facebook',
+                    location: 'LMS Dashboard'
```
s/LMS Dashboard/dashboard/

mulby commented on lms/static/js/course_sharing/course_sharing_events.js:
```diff
@@ -0,0 +1,28 @@
+        return function(courseId) {
+            $(".action-facebook[data-course-id='" + courseId + "']").on('click', function() {
+                // Emit an event telling that the Facebook share link was clicked.
+                Logger.log('edx.course.sharing.share_clicked', {
```
maybe just "edx.course.share_clicked"?

mulby replied:
also I suggest you move this logic to a single function that is called with the `social_media_site` as a parameter to reduce the chance of the two calls getting out of sync.

Qubad786 commented:
@mulby there were some configs went missing right after I updated the sandbox. I've fixed it. you may take a look now: https://course-sharing-on-lms-dashboard.sandbox.edx.org/dashboard

mulby commented:
@muzaffaryousaf I see it now, thanks!

When you finalize the event format, can you grab one of the events from the tracking logs and post it in this PR for our review?

Qubad786 commented:
@mulby sure, will do it in a sec.

Also, about UTM parameters, these are not added yet. can you provide an example( i.e. on where/how to use utm params) to have a sense of implementation in our case? I've been reading about it but I am unsure on what should those params contains? and should I just add them into URL as query params? Or need to handle those params someway in views too?

sstack22 commented:
@Qubad786  - just to clarify that the UTM parameter work is outside the scope of this PR. We can implement that after its live.

sstack22 commented:
@Qubad786 , @muzaffaryousaf - the share text suggested by marketing.

For Facebook: "I'm taking CourseName online with edX.org. Check it out!" (Link follows text)
For Twitter: "I'm taking CourseName online with @edxonline. Check it out!" (link follows text)

Qubad786 commented:
Twitter `share_clicked` event:
```
{  
   "username":"audit",
   "event_source":"browser",
   "name":"edx.course.share_clicked",
   "accept_language":"en-US,en;q=0.8",
   "time":"2017-04-12T18:45:13.582509+00:00",
   "agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36",
   "page":"http://localhost:8000/dashboard",
   "host":"precise64",
   "session":"126728b7a124666b8d40e455203f5464",
   "referer":"http://localhost:8000/dashboard",
   "context":{  
      "user_id":3,
      "org_id":"",
      "course_id":"",
      "path":"/event"
   },
   "ip":"10.0.2.2",
   "event":"{\"course_id\": \"course-v1:edX+DemoX+Demo_Course\", \"social_media_site\": \"twitter\", \"location\": \"dashboard\"}",
   "event_type":"edx.course.share_clicked"
}
```

Facebook `share_clicked` event:
```
{  
   "username":"audit",
   "event_source":"browser",
   "name":"edx.course.share_clicked",
   "accept_language":"en-US,en;q=0.8",
   "time":"2017-04-12T18:45:10.779385+00:00",
   "agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36",
   "page":"http://localhost:8000/dashboard",
   "host":"precise64",
   "session":"126728b7a124666b8d40e455203f5464",
   "referer":"http://localhost:8000/dashboard",
   "context":{  
      "user_id":3,
      "org_id":"",
      "course_id":"",
      "path":"/event"
   },
   "ip":"10.0.2.2",
   "event":"{\"course_id\": \"course-v1:edX+DemoX+Demo_Course\", \"social_media_site\": \"facebook\", \"location\": \"dashboard\"}",
   "event_type":"edx.course.share_clicked"
}
```

stroilova commented:
Thanks @mulby  and @muzaffaryousaf . This format looks good in general.

As is, they will be exported by course. cc @sstack22 -- is that OK, or did we want them to be internal only?

sstack22 commented:
@stroilova - I think it's fine to export by course. Thanks!

Qubad786 commented:
@sstack22 should these events go on Segment too? if so then I need to file an OPS ticket in order to add this event into the whitelist.

According to: http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/analytics.html#segment
> Additionally, the setting EVENT_TRACKING_SEGMENTIO_EMIT_WHITELIST in the lms.auth.json file can be used to specify event names that should be emitted to Segment from normal tracker.emit() calls. Events specified in this whitelist will be sent to both the tracking logs and Segment

Qubad786 commented:
@mulby Feedback is addressed :)

muzaffaryousaf commented:
@sstack22 Thanks for sharing text. We'll incorporate that soon!!

muzaffaryousaf commented:
@grantgoodmanedX Are you following along, this feature will land tomorrow probably. Let us know if you need any help. Thanks.

sstack22 commented:
@Qubad786 - yes, I think we want this to go to segment as well.

mulby commented:
Ah, so these events won't go to researchers because they are not emitted from a page whose URL contains the course_id. This is fine for now, but if/when we *do* want these to be visible to researchers we will have to make some changes to the javascript event collection infrastructure to support explicitly setting the course_id in the event context.

muzaffaryousaf commented:
@sstack22 The last note on this page, will be changed now http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/enable_socialsharing_icons.html?highlight=social%20sharing#configure-social-sharing
@grantgoodmanedX FYI ^^

Qubad786 commented:
@sstack22 please have look text has been accommodated.
I've updated sandbox: https://course-sharing-on-lms-dashboard.sandbox.edx.org/dashboard

grantgoodmanedX commented:
@muzaffaryousaf What is the change in the behavior for the note at the end of  http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/enable_socialsharing_icons.html ?
Can the note just be removed, because we'll be able to supply an appropriate URL?

muzaffaryousaf commented:
@grantgoodmanedX Now there will be value in course sharing url because of the change that we've made. Here is the order:
1 - social_sharing_url (if specified in advance settings)
2 - marketing_url
3 - link to LMS course about page

Qubad786 commented:
@sstack22 updated sandbox please take a look: https://course-sharing-on-lms-dashboard.sandbox.edx.org/dashboard

sstack22 commented:
@Qubad786 - looks great, 👍 !

grantgoodmanedX commented:
I have a PR for the doc changes: https://github.com/edx/edx-documentation/pull/1483

muzaffaryousaf commented on lms/templates/dashboard/_dashboard_course_listing.html:
```diff
@@ -449,6 +451,9 @@ <h4 class="message-title">${_('Your current verification will expire soon.')}</h
 </script>
+<%static:require_module module_name="js/course_sharing/course_sharing_events" class_name="CourseSharingEvents">
+    CourseSharingEvents("${course_overview.id | n, js_escaped_string}");
+</%static:require_module>
```
Let's re define this as:
```
<%block name="js_extra">
<%static:require_module module_name="js/course_sharing/course_sharing_events" class_name="CourseSharingEvents">
    CourseSharingEvents("${course_overview.id | n, js_escaped_string}");
</%static:require_module>
</%block>
```

For more details check `common/djangoapps/pipeline_mako/templates/static_content.html` file for  require_module and require_module_async.

muzaffaryousaf commented on lms/templates/dashboard/_dashboard_course_listing.html:
```diff
@@ -149,18 +150,15 @@ <h3 class="course-title">
-                  <% facebook_url = 'https://www.facebook.com/sharer/sharer.php?u=' + share_url %>
+                % if share_settings.get('DASHBOARD_FACEBOOK', False):
+                  <%
+                      share_text = _("I'm taking {course_name} online with edX.org. Check it out!").format(course_name=course_overview.display_name_with_default)
```
@sstack22 we're making this hardcode since this feature won't be enabled on white label sites. also, according to my little knowledge white label sites uses their own templates.

muzaffaryousaf commented on lms/templates/dashboard/_dashboard_course_listing.html:
```diff
@@ -169,15 +167,18 @@ <h3 class="course-title">
+                % if share_settings.get('DASHBOARD_TWITTER', False):
+                  <%
+                      default_share_text = _("I'm taking {course_name} online with @edxonline. Check it out!").format(course_name=course_overview.display_name_with_default)
+                      share_text = urllib.quote_plus(share_settings.get('DASHBOARD_TWITTER_TEXT', default_share_text.encode('utf-8')))
```
I would encode both values (DASHBOARD_TWITTER_TEXT and default_share_text) to avoid future error.
something like:
```
urllib.quote_plus(share_settings.get('DASHBOARD_TWITTER_TEXT', default_share_text).encode())

muzaffaryousaf commented on lms/templates/dashboard/_dashboard_course_listing.html:
```diff
@@ -169,15 +167,18 @@ <h3 class="course-title">
-                  <% twitter_url = 'https://twitter.com/intent/tweet?text=' + share_text + '%20' + share_url %>
+                % if share_settings.get('DASHBOARD_TWITTER', False):
+                  <%
+                      default_share_text = _("I'm taking {course_name} online with @edxonline. Check it out!").format(course_name=course_overview.display_name_with_default)
```
@sstack22 again making edxonline hard coded. Not a good practise though.

muzaffaryousaf commented:
@Qubad786 I'm done reviewing this PR and left couple of suggestions. Once you address them, don't forget to test the changes on your local as well as on sandbox and make sure js file is being served correctly.

muzaffaryousaf commented:
jenkins run js

muzaffaryousaf commented:
jenkins run bokchoy

muzaffaryousaf commented:
LGMT 👍

muhammad-ammar commented:
LGTM 👍
